File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/resources/library_analytics.py
"""Library Analytics — Figma REST API coverage.

Tier: enterprise only.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

if TYPE_CHECKING:
    from .._figma_client import FigmaClient


def _flatten(
    *,
    group_by: str,
    cursor: str | None = None,
    order: str | None = None,
    start_date: str | None = None,
    end_date: str | None = None,
) -> dict[str, Any]:
    params: dict[str, Any] = {"group_by": group_by}
    if cursor is not None:
        params["cursor"] = cursor
    if order is not None:
        params["order"] = order
    if start_date is not None:
        params["start_date"] = start_date
    if end_date is not None:
        params["end_date"] = end_date
    return params


class LibraryAnalyticsResource:
    def __init__(self, client: FigmaClient) -> None:
        self._client = client

    async def component_actions(self, file_key: str, *, group_by: str, **kwargs: Any) -> dict[str, Any]:
        return await self._client.get(
            f"/v1/analytics/libraries/{quote(file_key, safe='')}/component/actions",
            params=_flatten(group_by=group_by, **kwargs),
        )

    async def component_usages(self, file_key: str, *, group_by: str, **kwargs: Any) -> dict[str, Any]:
        return await self._client.get(
            f"/v1/analytics/libraries/{quote(file_key, safe='')}/component/usages",
            params=_flatten(group_by=group_by, **kwargs),
        )

    async def style_actions(self, file_key: str, *, group_by: str, **kwargs: Any) -> dict[str, Any]:
        return await self._client.get(
            f"/v1/analytics/libraries/{quote(file_key, safe='')}/style/actions",
            params=_flatten(group_by=group_by, **kwargs),
        )

    async def style_usages(self, file_key: str, *, group_by: str, **kwargs: Any) -> dict[str, Any]:
        return await self._client.get(
            f"/v1/analytics/libraries/{quote(file_key, safe='')}/style/usages",
            params=_flatten(group_by=group_by, **kwargs),
        )

    async def variable_actions(self, file_key: str, *, group_by: str, **kwargs: Any) -> dict[str, Any]:
        return await self._client.get(
            f"/v1/analytics/libraries/{quote(file_key, safe='')}/variable/actions",
            params=_flatten(group_by=group_by, **kwargs),
        )

    async def variable_usages(self, file_key: str, *, group_by: str, **kwargs: Any) -> dict[str, Any]:
        return await self._client.get(
            f"/v1/analytics/libraries/{quote(file_key, safe='')}/variable/usages",
            params=_flatten(group_by=group_by, **kwargs),
        )

File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/resources/library-analytics.ts
// @ts-nocheck
/**
 * Library Analytics — Figma REST API coverage.
 *
 * Tier: **enterprise only**. Returns 403 without an Enterprise/Org
 * plan. All endpoints are read-only and paginated via `cursor`.
 */

import type { FigmaClient } from '../figma-client.js';

export type AnalyticsGroupBy = 'component' | 'team' | 'file' | 'style' | 'variable';

export interface AnalyticsQueryParams {
  group_by: AnalyticsGroupBy;
  cursor?: string;
  order?: 'asc' | 'desc';
  // For date-bounded endpoints.
  start_date?: string;
  end_date?: string;
}

function flatten(params: AnalyticsQueryParams): Record<string, string> {
  const out: Record<string, string> = { group_by: params.group_by };
  if (params.cursor !== undefined) out.cursor = params.cursor;
  if (params.order !== undefined) out.order = params.order;
  if (params.start_date !== undefined) out.start_date = params.start_date;
  if (params.end_date !== undefined) out.end_date = params.end_date;
  return out;
}

export class LibraryAnalyticsResource {
  constructor(private readonly client: FigmaClient) {}

  /** GET /v1/analytics/libraries/:file_key/component/actions */
  async componentActions(fileKey: string, params: AnalyticsQueryParams): Promise<unknown> {
    return this.client.get(
      `/v1/analytics/libraries/${encodeURIComponent(fileKey)}/component/actions`,
      { params: flatten(params) },
    );
  }

  /** GET /v1/analytics/libraries/:file_key/component/usages */
  async componentUsages(fileKey: string, params: AnalyticsQueryParams): Promise<unknown> {
    return this.client.get(
      `/v1/analytics/libraries/${encodeURIComponent(fileKey)}/component/usages`,
      { params: flatten(params) },
    );
  }

  /** GET /v1/analytics/libraries/:file_key/style/actions */
  async styleActions(fileKey: string, params: AnalyticsQueryParams): Promise<unknown> {
    return this.client.get(`/v1/analytics/libraries/${encodeURIComponent(fileKey)}/style/actions`, {
      params: flatten(params),
    });
  }

  /** GET /v1/analytics/libraries/:file_key/style/usages */
  async styleUsages(fileKey: string, params: AnalyticsQueryParams): Promise<unknown> {
    return this.client.get(`/v1/analytics/libraries/${encodeURIComponent(fileKey)}/style/usages`, {
      params: flatten(params),
    });
  }

  /** GET /v1/analytics/libraries/:file_key/variable/actions */
  async variableActions(fileKey: string, params: AnalyticsQueryParams): Promise<unknown> {
    return this.client.get(
      `/v1/analytics/libraries/${encodeURIComponent(fileKey)}/variable/actions`,
      { params: flatten(params) },
    );
  }

  /** GET /v1/analytics/libraries/:file_key/variable/usages */
  async variableUsages(fileKey: string, params: AnalyticsQueryParams): Promise<unknown> {
    return this.client.get(
      `/v1/analytics/libraries/${encodeURIComponent(fileKey)}/variable/usages`,
      { params: flatten(params) },
    );
  }
}
